File: Chapter.03/FunctionalDataStructures.fs/LazyList.fs
module LazyList

// Listing 3.15 Lazy list implementation using F#
type LazyList<'a> =
    | Cons of head:'a * tail:Lazy<'a LazyList>  //#A
    | Empty
let empty = lazy(Empty)                         //#B

let rec append items list =                     //#C
    match items with
    | Cons(head, Lazy(tail)) ->
        Cons(head, lazy(append tail list))      //#D
    | Empty -> list

let list1 = Cons(42, lazy(Cons(21, empty)))     //#E
// val list1: LazyList<int> = Cons (42,Value is not created.)

let list = append (Cons(3, empty)) list1        //#F
// val list : LazyList<int> = Cons (3,Value is not created.)

let rec iter action list =                      //#G
    match list with
    | Cons(head, Lazy(tail)) ->
        action(head)
        iter action tail
    | Empty -> ()

list |> iter (printf "%d .. ")                  //#H
// 3 .. 42 .. 21 ..

File: Chapter.03/FunctionalDataStructures.cs/LazyList.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PersistentDataStructures
{
    // Listing 3.15 Lazy list implementation using C#
    public sealed class LazyList<T>
    {
        public LazyList(T head, Lazy<LazyList<T>> tail)
        {
            Head = head;
            Tail = tail;
            IsEmpty = false;
        }
        private LazyList()
        {
            IsEmpty = true;
        }
        public T Head { get; }
        public Lazy<LazyList<T>> Tail { get; }
        public bool IsEmpty { get; }

        public static readonly Lazy<LazyList<T>> Empty =
            new Lazy<LazyList<T>>(() => new LazyList<T>());
    }

    public static class LazyListExtensions
    {
        public static LazyList<T> Append<T>
            (this LazyList<T> list, LazyList<T> items)
        {
            if (items.IsEmpty) return list;
            return new LazyList<T>(items.Head,
                new Lazy<LazyList<T>>(() =>
                            list.Append(items.Tail.Value)));
        }

        public static void Iterate<T>
            (this LazyList<T> list, Action<T> action)
        {
            if (list.IsEmpty)
                return;
            action(list.Head);
            list.Tail.Value.Iterate(action);
        }
    }
}
